src/server.cpp

```diff
@@ -168,138 +171,216 @@ void installSignalHandlers() {
 
+public:
+    ProfilerModule() = default;
+    int start(const Config& config) override {
 #ifdef MTR_ENABLED
-        auto profiler = Profiler(config.tracePath());
+        state = ModuleState::STARTED_INITIALIZE;
+        auto profiler = std::make_unique<Profiler>(config.tracePath());
         if (!profiler.isInitialized()) {
             SPDLOG_ERROR("Cannot open file for profiler, --trace_path: {}", config.tracePath());
             return EXIT_FAILURE;
         }
+        state = ModuleState::INITIALIZED;
 #endif
+        return EXIT_SUCCESS;
+    }
+    void shutdown() override {
+        state = ModuleState::STARTED_SHUTDOWN;
+        profiler.reset();
+        state = ModuleState::SHUTDOWN;
+    }
+};
 
-        PredictionServiceImpl predict_service;
-        ModelServiceImpl model_service;
-        KFSInferenceServiceImpl kfsGrpcInferenceService;
-
-        auto grpc = startGRPCServer(predict_service, model_service, kfsGrpcInferenceService);
-        auto rest = startRESTServer();
+void Server::setShutdownRequest(int i) {
+    shutdownRequest = 1;
+    shutdown_request = 1;
+}
 
+int Server::start(int argc, char** argv) {
+    ovms::Server& server = ovms::Server::instance();
+    installSignalHandlers(server);
+    auto retCode = EXIT_SUCCESS;
+    try {
+        auto& config = ovms::Config::instance().parse(argc, argv);
+        configure_logger(config.logLevel(), config.logPath());
+#ifdef MTR_ENABLED
+        this->modules.emplace("ProfilerModule", std::make_unique<ProfilerModule>());
+        retCode = modules.at("ProfilerModule")->start(config);
+        if (retCode)
+            return retCode;
+#endif
+        this->modules.emplace(std::string("GRPCServerModule"), std::make_unique<GRPCServerModule>());
+        retCode = modules.at("GRPCServerModule")->start(config);
+        if (retCode)
+            return retCode;
+        this->modules.emplace("HTTPServerModule", std::make_unique<HTTPServerModule>());
+        retCode = modules.at("HTTPServerModule")->start(config);
+        if (retCode)
+            return retCode;
+        this->modules.emplace("ServableManagerModule", std::make_unique<ServableManagerModule>());
+        retCode = modules.at("ServableManagerModule")->start(config);
+        if (retCode)
+            return retCode;
         while (!shutdown_request) {
             std::this_thread::sleep_for(std::chrono::seconds(1));
         }
         if (shutdown_request == 2) {
             SPDLOG_ERROR("Illegal operation. OVMS started on unsupported device");
         }
         SPDLOG_INFO("Shutting down");
-        for (const auto& g : grpc) {
-            g->Shutdown();
-            SPDLOG_INFO("Shutdown gRPC server");
-        }
-
-        if (rest != nullptr) {
-            rest->Terminate();
-            rest->WaitForTermination();
-            SPDLOG_INFO("Shutdown HTTP server");
-        }
-
-        ModelManager::getInstance().join();
+        modules.at("GRPCServerModule")->shutdown();
```

Comment:
Could shutdown be called automatically during module destructor?

Reply:
You want to keep modules as a Guard helper inside modules container and just reset those?

I wanted to keep the order of starting/shutting down each component - eg. gRPC server won't stop until all requests are handled. This forces us to have specific order of closing. We could still save the order with just removing specific module.

So the answer is yes we could but I am not sure if we want.